### ISSUE TITLE
Is there any way to define what is being returned from the service (wrapping it into a class)?

### ISSUE THREAD

**Issue by caiobiodere** [KnisterPeter/pretend#515]
Hey everyone first of all thanks for creating this, I was thinking about doing something like this for a project but you guys got it covered! second here comes my question I still don't know if it is possible to define a response type like a class being created from what is returned from the service. I would probably also want to define which path of the object should be pointed to this class this could be a decorator. What do you guys think about all of this? I might be interested in helping this project it is really what I have been doing with Swift using Sourcery :)

**Comment by KnisterPeter** [comment]
Hi @caiobiodere, 
I'm not sure if I understand what you are aiming at. What do you mean by 'return like a typescript class'?
Do you want to have the returned data wrapped in a class which is created based on the method which was called?

**Comment by caiobiodere** [comment]
@KnisterPeter yeah that's exactly what I meant sorry english is not one of my best things yet, still getting in there :) yes sure can do that give me a couple of days will figure out how to do that

**Comment by KnisterPeter** [comment]
@caiobiodere no need to say sorry :smiley: We all are non native speakers. You're welcome.

Before start coding a rough scribble of pseudo code might help me to understand your requirements. Maybe what you try to achieve is already possible, not sure.
Please just explain it here in a few more sentences and a sketch.

**Comment by caiobiodere** [comment]
@KnisterPeter thanks for understanding mate!

getting into it, the idea behind all of this would be something similar to this:

type UserType {
}

class User: UserType {
}

```
@Headers('Accept: application/json')
@Get('/user/{id}', true)
@Response(type: User | User[], keyPath: "userObject" |  null) (the new decorator I mentioned before)
public async get(id: string, parameters: any) : User | User[] {}
```

**Comment by KnisterPeter** [comment]
Hi @caiobiodere,
this looks nice. Can you explain a bit what you need the `keyPath` for? The `type` is pretty obvious.

In general I could think of adding this feature, I like it.

**Comment by caiobiodere** [comment]
Hi mate @KnisterPeter,
basically keyPath would be the path that it should extract the object to be wrapped from, lets say that we have a response like this:
```
{
  user: {
    id: 1
  }
}
```

I could use keyPath as "user" and tell the decorator to wrap into the User object from it, this would save a lot of scenarios I can tell that because most of the time API's are not in the way you really want them to be

**Comment by KnisterPeter** [comment]
Instead I would propose to allow this to be a function. Then if the reponse is not matching it could be updated by that. Like this:

```ts
@Response(type: User, transform: (input) => input.user)
public async get(id: string, parameters: any) : User {}
```

**Comment by caiobiodere** [comment]
I like your idea better, do you want me to help with that? I really want this to be part of my current working project

**Comment by KnisterPeter** [comment]
You can give it a try. I would accept a PR if you like.
But I'm as well a bit biased and have an idea how I would implement this. Lets see, maybe I like your idea as well. :smiley:

**Comment by caiobiodere** [comment]
Lets get this running, will be watching and forking this project from now on

**Comment by caiobiodere** [comment]
if you think you can finish this in a couple of days I can wait mate, that would be no problem at all

**Comment by KnisterPeter** [comment]
@caiobiodere I'll definitely build this this week I think.

**Comment by caiobiodere** [comment]
@KnisterPeter hey mate I saw you started it already, how is it going do you have an estimated time?

**Comment by KnisterPeter** [comment]
The PR could be reviewed by you, it's already done.
The tests should hold some examples.
Please give some feedback.

**Comment by KnisterPeter** [comment]
@caiobiodere Any comments or thoughts? I'd like to make a release out of this.

**Comment by KnisterPeter** [comment]
These are the docs in the [README](https://github.com/KnisterPeter/pretend/tree/data-mapper#data-mappers)

**Comment by KnisterPeter** [comment]
Released in v3.1.0

**Comment by caiobiodere** [comment]
thanks mate! I think that's good enough regarding it also fixes my case :)
